File: 12-项目工程/01-数据结构与集合选型/GoMap可扩展散列与SwissTable/go/main.go
package main

import "fmt"

const golden = uint64(0x9E3779B97F4A7C15)

func mix(i int) uint64 { return uint64(i) * golden }

func main() {
	fmt.Println("== 常量 ==")
	fmt.Printf("  MapGroupSlots=%d maxAvgGroupLoad=%d maxTableCapacity=%d\n",
		mapGroupSlots, maxAvgGroupLoad, maxTableCapacity)
	fmt.Printf("  ctrlEmpty=0x%02X ctrlDeleted=0x%02X\n", ctrlEmpty, ctrlDeleted)

	fmt.Println("\n== h1 / h2 ==")
	for _, h := range []uint64{0x00, 0x7F, 0x80, 0xFF, 0x1FF, ^uint64(0)} {
		fmt.Printf("  hash=0x%016X -> h1=%d h2=0x%02X\n", h, h1(h), h2(h))
	}

	fmt.Println("\n== 单表容量与 growthLeft ==")
	for _, c := range []int{8, 16, 64, 256, 1024} {
		fmt.Printf("  capacity=%-6d growthLeft=%-6d (%.3f)\n",
			c, maxGrowthLeft(c), float64(maxGrowthLeft(c))/float64(c))
	}

	fmt.Println("\n== NewMap(hint) 的目录推导 ==")
	for _, hint := range []int{8, 9, 100, 1000, 10000, 100000} {
		lay := newMapLayout(hint)
		if lay.small {
			fmt.Printf("  hint=%-8d 单组小图（dirLen=0，8 槽全可用）\n", hint)
		} else {
			fmt.Printf("  hint=%-8d target=%-8d dirLen=%-4d globalDepth=%d globalShift=%d 每表 %d 槽\n",
				hint, lay.targetCapacity, lay.dirLen, lay.globalDepth, lay.globalShift, lay.tableCapacity)
		}
	}

	fmt.Println("\n== 三角探测（组下标）==")
	for _, groups := range []int{8, 16} {
		fmt.Printf("  组数 %3d 从 h1=0 起: %v\n", groups, probeGroups(0, groups-1, 8))
	}

	fmt.Println("\n== 生命周期：小图 -> grow -> split -> 目录翻倍 ==")
	m := newGoMap(0)
	prevCap, prevDir, prevDepth := mapGroupSlots, 0, 0
	for i := 1; i <= 2600; i++ {
		m.put(fmt.Sprintf("k%d", i), mix(i))
		cap, dirLen, depth := m.totalCapacity(), len(m.dir), m.globalDepth
		if m.small {
			cap, dirLen = mapGroupSlots, 0
		}
		if cap != prevCap || dirLen != prevDir || depth != prevDepth {
			fmt.Printf("  第 %4d 个 capacity=%-5d dirLen=%-4d globalDepth=%d 表数=%d\n",
				i, cap, dirLen, depth, len(m.tables()))
			prevCap, prevDir, prevDepth = cap, dirLen, depth
		}
	}
	fmt.Printf("  splits=%d 目录翻倍=%d grows=%d\n", m.splits, m.dirDoublings, m.grows)

	fmt.Println("\n== 墓碑：只有『整组都满』时删除才产生墓碑 ==")
	t1 := newTable(8, 0, 0)
	for i := 1; i <= 7; i++ {
		t1.uncheckedPut(fmt.Sprintf("g%d", i), uint64(i)*0x400)
	}
	t1.delete("g1", 0x400)
	fmt.Printf("  单组表(8槽) 删 1 个 -> 墓碑=%d growthLeft=%d（单组表恒留空槽，永不产生墓碑）\n",
		t1.tombstones(), t1.growthLeft)

	t2 := newTable(16, 0, 0)
	for j := 1; j <= 8; j++ {
		t2.uncheckedPut(fmt.Sprintf("h%d", j), uint64(j)) // h1 = 0，全落组 0
	}
	gl := t2.growthLeft
	t2.delete("h1", 1)
	fmt.Printf("  16 槽表 组0 填满后删 1 个 -> 墓碑=%d growthLeft %d -> %d\n",
		t2.tombstones(), gl, t2.growthLeft)
	t2.put("new", 0x7F01)
	fmt.Printf("  再插 1 个复用墓碑 -> 墓碑=%d growthLeft=%d\n", t2.tombstones(), t2.growthLeft)

	fmt.Println("\n== 1024 槽表满了之后：split 而不是继续翻倍 ==")
	t3 := newTable(1024, 0, 3)
	gm := &goMap{globalDepth: 3, dir: []*table{t3}}
	for i := 0; i < 896; i++ {
		t3.uncheckedPut(fmt.Sprintf("f%d", i), mix(i))
	}
	fmt.Printf("  拆分前: dirLen=%d globalDepth=%d capacity=%d used=%d growthLeft=%d\n",
		len(gm.dir), gm.globalDepth, gm.totalCapacity(), t3.used, t3.growthLeft)
	t3.rehash(gm)
	fmt.Printf("  拆分后: dirLen=%d globalDepth=%d capacity=%d 表数=%d splits=%d 目录翻倍=%d\n",
		len(gm.dir), gm.globalDepth, gm.totalCapacity(), len(gm.tables()), gm.splits, gm.dirDoublings)
	fmt.Printf("  分裂位 localDepthMask(4) = 0x%X\n", localDepthMask(4))
}

File: 12-项目工程/01-数据结构与集合选型/GoMap可扩展散列与SwissTable/python/main.py
"""662 演示入口：小图快路径 → grow → split → 目录翻倍的全过程。"""

from __future__ import annotations

from go_swissmap import (
    CTRL_DELETED,
    CTRL_EMPTY,
    MAP_GROUP_SLOTS,
    MAX_AVG_GROUP_LOAD,
    MAX_TABLE_CAPACITY,
    GoMap,
    Table,
    align_up_pow2,
    h1,
    h2,
    local_depth_mask,
    max_growth_left,
    new_map_layout,
    new_table_capacity,
    probe_groups,
)

GOLDEN = 0x9E3779B97F4A7C15
U64 = (1 << 64) - 1


def mix(i: int) -> int:
    """确定性伪随机：用黄金比例常数打散，避免断言受随机源影响。"""
    return (i * GOLDEN) & U64


def show_constants() -> None:
    print("== 常量 ==")
    print(f"  MapGroupSlots={MAP_GROUP_SLOTS}  maxAvgGroupLoad={MAX_AVG_GROUP_LOAD}"
          f"  maxTableCapacity={MAX_TABLE_CAPACITY}")
    print(f"  ctrlEmpty=0x{CTRL_EMPTY:02X}  ctrlDeleted=0x{CTRL_DELETED:02X}")
    print("  （Go 与 Abseil 用同一组哨兵；Rust hashbrown 是反过来的 0xFF / 0x80，见 663）")


def show_hash_split() -> None:
    print("\n== h1 / h2 ==")
    for h in (0x00, 0x7F, 0x80, 0xFF, 0x1FF, U64):
        print(f"  hash=0x{h:016X} -> h1=0x{h1(h):014X} ({h1(h)})  h2=0x{h2(h):02X}")


def show_capacity() -> None:
    print("\n== 单表容量与 growthLeft ==")
    for cap_ in (8, 16, 64, 256, 1024):
        print(f"  capacity={cap_:<6} growthLeft={max_growth_left(cap_):<6}"
              f"  ({max_growth_left(cap_)}/{cap_} = {max_growth_left(cap_) / cap_:.3f})")
    print(f"  newTable: 1 -> {new_table_capacity(1)}, 9 -> {new_table_capacity(9)},"
          f" 100 -> {new_table_capacity(100)}, 700 -> {new_table_capacity(700)}")


def show_new_map() -> None:
    print("\n== NewMap(hint) 的目录推导 ==")
    for hint in (8, 9, 100, 1000, 10000, 100000):
        lay = new_map_layout(hint)
        if lay["small"]:
            print(f"  hint={hint:<8} 单组小图（dirLen=0，8 槽全可用）")
        else:
            print(f"  hint={hint:<8} target={lay['target_capacity']:<8} dirLen={lay['dir_len']:<4}"
                  f" globalDepth={lay['global_depth']} globalShift={lay['global_shift']}"
                  f" 每表 {lay['table_capacity']} 槽")


def show_probe() -> None:
    print("\n== 三角探测（组下标）==")
    for groups in (8, 16):
        print(f"  组数 {groups:>3} 从 h1=0 起:", list(probe_groups(0, groups - 1, 8)))
    print("  源码 next(): index++; offset = (offset + index) & mask —— 步长 1,2,3,4,…")


def show_lifecycle() -> None:
    print("\n== 一次完整生命周期：小图 -> grow -> split -> 目录翻倍 ==")
    m = GoMap(0)
    prev = m.snapshot()
    print(f"  start   {prev}")
    for i in range(1, 2600):
        m.put(f"k{i}", mix(i))
        s = m.snapshot()
        if (s["capacity"], s["dir_len"], s["global_depth"]) != (
            prev["capacity"], prev["dir_len"], prev["global_depth"]
        ):
            print(f"  第 {i:>4} 个 {s}")
            prev = s
    print(f"  终态    {m.snapshot()}")
    print(f"  localDepthMask(globalDepth) = 1 << {64 - m.global_depth}")


def show_tombstone() -> None:
    print("\n== 墓碑：只有『整组都满』时删除才产生墓碑 ==")
    t = Table(8, 0, 0)
    for i in range(7):
        t.unchecked_put(f"g{i}", (i + 1) * 0x400)
    t.delete("g0", 0x400)
    print(f"  单组表(8槽) 删 1 个 -> 墓碑={t.tombstones()} growthLeft={t.growth_left}"
          "   （单组表恒留一个空槽，永不产生墓碑）")
    t2 = Table(16, 0, 0)
    for j in range(8):
        t2.unchecked_put(f"h{j}", j + 1)      # h1 = 0，全落组 0
    gl = t2.growth_left
    t2.delete("h0", 1)
    print(f"  16 槽表 组0 填满后删 1 个 -> 墓碑={t2.tombstones()} "
          f"growthLeft {gl} -> {t2.growth_left}   （墓碑不归还配额）")
    t2.put("new", 0x7F01)
    print(f"  再插 1 个复用墓碑 -> 墓碑={t2.tombstones()} growthLeft={t2.growth_left}")


def show_split() -> None:
    print("\n== 1024 槽表满了之后：split 而不是继续翻倍 ==")
    t = Table(1024, 0, 3)
    gm = GoMap(0)
    gm.small = False
    gm.dir = [t]
    gm.global_depth = 3
    for i in range(896):
        t.unchecked_put(f"f{i}", mix(i))
    print(f"  拆分前: dirLen={len(gm.dir)} globalDepth={gm.global_depth} "
          f"capacity={gm.total_capacity()} used={t.used} growthLeft={t.growth_left}")
    t.rehash(gm)
    print(f"  拆分后: dirLen={len(gm.dir)} globalDepth={gm.global_depth} "
          f"capacity={gm.total_capacity()} 表数={len(gm.tables())} "
          f"splits={gm.splits} 目录翻倍={gm.dir_doublings}")
    print(f"  分裂位 localDepthMask(4) = 1 << {64 - 4} = 0x{local_depth_mask(4):X}")


if __name__ == "__main__":
    show_constants()
    show_hash_split()
    show_capacity()
    show_new_map()
    show_probe()
    show_lifecycle()
    show_tombstone()
    show_split()
